processor/src/main/java/com/linecorp/decaton/processor/SubscriptionStateListener.java

```diff
@@ -0,0 +1,108 @@
+ * under the License.
+ */
+
+package com.linecorp.decaton.processor;
+
+import com.linecorp.decaton.processor.runtime.ProcessorSubscription;
+
+/**
+ * An interface to listen state changes of {@link ProcessorSubscription}.
+ * This interface is intended to be used for health check or integration testing purpose.
+ *
+ * For example, you can make sure that Decaton has started to process as follows:
+ * <pre>
+ * {@code
+ * CountDownLatch latch = new CountDownLatch(1);
+ * SubscriptionBuilder.newBuilder("my-subscription")
+ *                    // ...
+ *                    .stateListener(state -> {
+ *                        if (state == State.RUNNING) {
+ *                            latch.countDown();
+ *                        }
+ *                    }).buildAndStart();
+ * latch.await();
+ * // actual integration testing continues...
+ * }
+ * </pre>
+ */
+public interface SubscriptionStateListener {
+    /**
+     * Represents possible states that a {@link ProcessorSubscription} can be in.
+     * The expected state transition is:
+     *
+     * <pre>
+     *
+     *                 +-----+--------+
+     *                 | INITIALIZING |
+     *                 +-----+--------+
+     *                       |
+     *                       v
+     *                 +--------------+
+     *                 | REBALANCING  |
+     *                 +----+--+------+
+     *                      |  ^
+     *                      v  |
+     *                 +----+--+------+
+     *                 |   RUNNING    |
+     *                 +-----+--------+
+     *                       |
+     *                       v
+     *                 +-----+--------+
+     *                 |SHUTTING_DOWN |
+     *                 +-----+--------+
+     *                       |
+     *                       v
+     *                 +-----+--------+
+     *                 | TERMINATED   |
+     *                 +--------------+
+     *
+     * </pre>
+     * Listener will be called at each transition in the flow.
+     * In addition, listener will be also called with INITIALIZING state when Decaton starts startup sequence
+     */
+    enum State {
+        /**
+         * Initializing subscription internals.
+         */
+        INITIALIZING,
+        /**
+         * Consumer group rebalance is running.
+         * Meanwhile no extra records are fetched until it completes.
+         */
+        REBALANCING,
+        /**
+         * Started to process tasks.
+         * Fetching records from brokers and feeding them into partition processors.
+         */
+        RUNNING,
+        /**
+         * Entered shutdown sequence.
+         * No extra tasks will be queued, but existing queued tasks will be processed.
+         */
+        SHUTTING_DOWN,
+        /**
+         * All shutdown sequence has done. The subscription is terminated completely.
+         */
+        TERMINATED,
+    }
+
+    /**
+     * Called after state transitioned to new state.
```

Comment:
`after` is kind of ambiguous as in there's no associated "work" for each states? rather they're triggered at between two works. so "at"?